kyo-kernel/shared/src/main/scala/kyo/kernel/internal/CanLift.scala

```diff
@@ -0,0 +1,64 @@
+package kyo.kernel.internal
+
+import kyo.<
+import scala.annotation.implicitNotFound
+import scala.quoted.*
+import scala.util.NotGiven
+
+/** CanLift is a "soft" constraint that indicates a type should not contain nested effect computations (A < S), or A is not a module from
+  * kyo (like Abort.type).
+  *
+  * This constraint helps:
+  *  - prevent accidental nesting of effects that would require flattening, but cannot be strictly enforced in all generic contexts,
+  *  - prevent calling combinators from (A < S) on modules, like Abort.foldAbort.
+  *
+  * @tparam A
+  *   The type to check for nested effects
+  */
+@implicitNotFound("""
+Type '${A}' may contain a nested effect computation.
+
+This usually means you have a value of type 'X < S' where you need a plain value.
+To fix this, you can:
+
+1. Call .flatten to combine the nested effects:
+     (x: (Int < S1) < S2).flatten  // Result: Int < (S1 & S2)
+
+2. Split the computation into multiple statements:
+     val x: Int < S1 = computeFirst()
+     val y: Int < S2 = useResult(x)
+""")
+opaque type CanLift[A] = Null
+
+object CanLiftMacro:
+    inline given derived[A](using inline ng: NotGiven[A <:< (Any < Nothing)]): CanLift[A] = ${ liftImpl[A] }
```

Comment:
circular references don't help with the macro:
- `NotGiven[A <:< (Any < Nothing)]` is better than doing it in the macro.
- the macro have to be isolated from `inline given CanLift[Nothing]`, hence the `CanLiftMacro`, then an export.